icechunk/src/ops/stats.rs

```diff
@@ -10,37 +11,138 @@ use tracing::trace;
+            ))
+        })?
+        .insert(key)
+    {
+        *size_counter += size_increment;
+    }
+    Ok(())
+}
+
 fn calculate_manifest_storage(
     manifest: Arc<Manifest>,
-    seen_chunks: Arc<Mutex<HashSet<ChunkId>>>,
-) -> RepositoryResult<u64> {
+    // Different types of chunks require using different types of ids to de-duplicate them when counting.
+    seen_native_chunks: Arc<Mutex<HashSet<ChunkId>>>,
+    seen_virtual_checksums: Arc<Mutex<HashSet<Checksum>>>,
+    // Virtual chunks don't necessarily have checksums. For those which don't we instead use the (url, offset, length) tuple as an identifier.
+    // This is more expensive, but should still work to de-duplicate because the only way that this identifier could be the same for different chunks
+    // is if the data were entirely overwritten at that exact storage location. In this scenario it makes sense not to count both chunks towards the storage total,
+    // as the overwritten data is no longer accessible anyway.
+    seen_virtual_identifiers: Arc<
+        Mutex<HashSet<(VirtualChunkLocation, ChunkOffset, ChunkLength)>>,
+    >,
+) -> RepositoryResult<ChunkStorageStats> {
     trace!(manifest_id = %manifest.id(), "Processing manifest");
-    let mut size = 0;
+    let mut native_bytes: u64 = 0;
+    let mut virtual_bytes: u64 = 0;
+    let mut inlined_bytes: u64 = 0;
     for payload in manifest.chunk_payloads() {
         match payload {
             Ok(ChunkPayload::Ref(chunk_ref)) => {
-                if seen_chunks
-                    .lock()
-                    .map_err(|e| {
-                        RepositoryErrorKind::Other(format!(
-                            "Thread panic during manifest_chunk_storage: {e}"
-                        ))
-                    })?
-                    .insert(chunk_ref.id)
-                {
-                    size += chunk_ref.length;
+                // Deduplicate native chunks by ChunkId
+                insert_and_increment_size_if_new(
+                    &seen_native_chunks,
+                    chunk_ref.id,
+                    chunk_ref.length,
+                    &mut native_bytes,
+                )?;
+            }
+            Ok(ChunkPayload::Virtual(virtual_ref)) => {
+                // Deduplicate by checksum if available, otherwise by (location, offset, length)
+                if let Some(checksum) = &virtual_ref.checksum {
+                    // Has checksum: deduplicate by checksum
+                    insert_and_increment_size_if_new(
+                        &seen_virtual_checksums,
+                        checksum.clone(),
+                        virtual_ref.length,
+                        &mut virtual_bytes,
+                    )?;
+                } else {
+                    // No checksum: deduplicate by (location, offset, length)
+                    let virtual_identifier = (
+                        virtual_ref.location.clone(),
```

Comment:
FYI I think it's possible to remove these clones (by passing a reference and then only actually cloning if needed), but it comes at the cost of an additional if. Maybe worth it though to minimize memory usage?